### ISSUE TITLE
url and link types

### ISSUE THREAD

**Issue by antoniogarrote** [raml-org/raml-spec#540]
This is a proposal to include new primitive data types into RAML types to describe payloads including urls and links.

This is an example payload from Github's API:

``` json
{
  "login": "octocat",
  "id": 1,
  "avatar_url": "https://github.com/images/error/octocat_happy.gif",
  "gravatar_id": "",
  "url": "https://api.github.com/users/octocat",
  "html_url": "https://github.com/octocat",
  "followers_url": "https://api.github.com/users/octocat/followers",
  "following_url": "https://api.github.com/users/octocat/following{/other_user}",
  "gists_url": "https://api.github.com/users/octocat/gists{/gist_id}",
  "starred_url": "https://api.github.com/users/octocat/starred{/owner}{/repo}",
  "subscriptions_url": "https://api.github.com/users/octocat/subscriptions",
  "organizations_url": "https://api.github.com/users/octocat/orgs",
  "repos_url": "https://api.github.com/users/octocat/repos",
  "events_url": "https://api.github.com/users/octocat/events{/privacy}",
  "received_events_url": "https://api.github.com/users/octocat/received_events",
  "type": "User",
  "site_admin": false,
  "name": "monalisa octocat",
  "company": "GitHub",
  "blog": "https://github.com/blog",
  "location": "San Francisco",
  "email": "octocat@github.com",
  "hireable": false,
  "bio": "There once was...",
  "public_repos": 2,
  "public_gists": 1,
  "followers": 20,
  "following": 0,
  "created_at": "2008-01-14T04:33:35Z",
  "updated_at": "2008-01-14T04:33:35Z"
}
```

Properties like `url`, `html_url`, `followers_url`, etc contain as a value URLs that can be followed by a HTTP agent to discover new resources.
This is a feature common in RESTful APIs using hypermedia.

Right now the only option for describing these properties in RAML is to use the `string` type.
Is it always possible to add a custom annotation to mark the string as a URL, but since URLs are one of the most basic building blocks in the web, it might be possible to standardize URLs with its own type or facet.

The second part of this issue is about links.

The github payload is using plain JSON to encode the information. JSON is not aware of the concept of URL or link between resources using a URL so it does not provide syntax to mark a property with a URL as a link, in the same way HTML uses the combination of the `a` tag and the `href` attribute.
Some serialisation formats like JSON-Schema or JSON-LD, designed that are aware of the notion of link define, their own syntax for expressing links:

``` json
{
  "foaf:knows": { "@id": "http://twitter.com/joearms" }
}
```

When describing payloads in these formats in RAML it would be useful to mark that syntax as `link` type inheriting from `object` wrapping a URL:

``` raml
properties:
  "foaf:knows":
    type: link
    ref: "@id"
```

Where the facet `ref` will mean that the property `@id` is of type `url`.

The same type could be used to describe in RAML other hypermedia aware formats, for example HAL:

``` json
{
    "_links": {
        "self": {
            "href": "http://example.org/api/user/matthew"
        }
    },
    "id": "matthew",
    "name": "Matthew Weier O'Phinney"
}
```

``` raml
properties:
  self?:
    type: link
    ref: href
```

Having both types, or at least the `url` type, will make it easier for RAML to describe REST APIs supporting hypermedia.

**Comment by sichvoge** [comment]
I agree to the `url` type since it makes sense but not really sure about the `ref` type since it would tie HTTP concepts with RAML data types. Any comments?

**Comment by petrochenko-pavel-a** [comment]
As for me  I do not see any need to include this types into set of the built in types in the language, but of course I am fully agree that they are usefull and should be part of "standard" set of libraries 

Regards,
Pavel

**Comment by ageoghegan** [comment]
Great requirement.  Trying to be comprehensive and apply a URL regular expression pattern was just too difficult.  I'd prefer some flexibility in what constitutes acceptable URLs though, scheme's, ports etc. would be nice.
